### PR TITLE
update licences

### PR DESCRIPTION
**Pull request by SylvainJuge**
- **licences update**
- **use version catalog**
- **generate a single file notice file**
- **embed asm bsd 3 clause license**
- **embed MPL 2 license**
- **include NOTICE and LICENSE in libraries (jar) and agent (jar)**

re-opens closed PR due to making the project public: #128 

## Implementation details:

In addition of the project License (Apache2), two extra licenses are included to cover dependencies (they are linked from the `/NOTICE` file)
- `/licenses/LICENSE_asm_bsd-3-clause` 
- `/licenses/LICENSE_mpl-2`

In source code:

- the `/LICENSE` contains a copy of the Apache 2 license
- the `/NOTICE` is generated from a template
- the `/licenses/` folder contains:
  - one `more-licenses.md` file that contains the list of dependencies and their licences with links to sub-folders
  - one sub-folder per embedded dependency of the project with license and notice if they are packaged in the dependency
  
In packaged artifacts (in particular for the java agent):

- the `/LICENSE` file is copied to `/META-INF/LICENSE`
- the `/NOTICE` file is copied to `/META-INF/NOTICE`
- the `/licenses` folder from the upstream agent is copied as-is to `/META-INF/licenses`
- the `/licenses` folder content is copied to `/META-INF/licenses`, hence it complements what is included in the upstream agent.
- two markdown files describe the dependencies `/META-INF/licenses/more-licenses.md` (from this project) and `/META-INF/licenses/licenses.md` (from the upstream agent), some overlap is expected but there is no harm for duplication here.

To update:

- the `update-license-and-notice` gradle task updates the `/NOTICE` and `/licenses/more-licenses.md` + sub-folders.

### CODE REVIEW

File: NOTICE [hunk 1]
```diff
@@ -1,6 +1,2551 @@
...
+          Copyright 2023-2024 Elasticsearch B.V.
+
+This project is licensed under the Apache License, Version 2.0 - https://www.apache.org/licenses/LICENSE-2.0
+An embedded copy of the Apache License, Version 2.0 is provided below.
+
+This project depends on ASM which is licensed under the 3-Clause BSD License - https://opensource.org/licenses/BSD-3-Clause
+An embedded copy of the ASM license (3-Clause BSD License) is provided below.
+
+This project depends on okhttp, which contains 'publicsuffixes.gz' that is licensed under the Mozilla Public License, v. 2.0 - https://mozilla.org/MPL/2.0/
+An embedded copy of the publicsuffixes.gz license (Mozilla Public License, v. 2.0) is provided below.
```
Comment (SylvainJuge) [inline NOTICE:11]: [for reviewer] the complete notice file is now generated, only the first few lines are manually added in the `build.gradle` file.

File: build.gradle [hunk 2]
```diff
@@ -1,11 +1,14 @@
...
+
+import java.nio.file.Files
+import java.nio.file.StandardOpenOption
 
 plugins {
   id "idea"
   id "maven-publish"
   id "base" // required to get 'clean' task
-  id 'com.github.jk1.dependency-license-report' version '2.5'
+  alias(catalog.plugins.licenseReport)
```
Comment (SylvainJuge) [inline build.gradle:11]: [for reviewer] just use version catalog

File: build.gradle [hunk 3]
```diff
@@ -142,12 +145,52 @@ tasks.configure {
...
+
+  lines.addAll(Arrays.asList("","###### publicsuffixes.gz Mozilla Public License, v. 2.0", ""))
+  lines.addAll(Files.readAllLines(licences.resolve("publicsuffixes_mpl-2")))
+
+  lines.addAll(Arrays.asList("","###############################################################################"))
+
+  lines.addAll(Files.readAllLines(licences.resolve("NOTICE_dependencies")))
+
+  Files.write(rootDir.toPath().resolve("NOTICE"), lines)
 }
```
Comment (SylvainJuge) [inline build.gradle:189]: [for reviewer] this generates the content of the `NOTICE` file at the root of the project, which is then embedded as-is in the produced binaries.

File: build.gradle [hunk 4]
```diff
@@ -142,12 +145,52 @@ tasks.configure {
...
 
 licenseReport {
   outputDir = "$projectDir/licenses"
-  renderers = [new InventoryMarkdownReportRenderer("licenses.md")]
+  renderers = [new TextReportRenderer("NOTICE_dependencies")]
```
Comment (SylvainJuge) [inline build.gradle:193]: [for reviewer] writes dependencies into an intermediary checked-in file that allows to review dependency changes when they happen.

File: licenses/asm_bsd-3-clause [hunk 5]
```diff
@@ -0,0 +1,28 @@
+
```
Comment (SylvainJuge) [inline licenses/asm_bsd-3-clause:1]: [for reviewer] we need to embed a copy of the ASM BSD 3 clause license.

File: licenses/publicsuffixes_mpl-2 [hunk 6]
```diff
@@ -0,0 +1,373 @@
+Mozilla Public License Version 2.0
```
Comment (SylvainJuge) [inline licenses/publicsuffixes_mpl-2:1]: [for reviewer] we need to embed a copy of the MPL 2.0 license as it's included through okhttp dependency.

**Review by jackshirazi — APPROVED** [review]

**Review by jackshirazi — APPROVED** [review]

**Pull request merged**